File: packages/python/speech_sdk/providers/openai.py
"""OpenAI TTS provider — mirrors src/providers/openai/."""

from __future__ import annotations

from dataclasses import dataclass

import httpx

from ..provider_utils import handle_error_response, resolve_api_key
from ..speech_provider import SpeechRequest
from ..speech_result import GeneratedAudioFile, SpeechResult

_DEFAULT_BASE_URL = "https://api.openai.com/v1"
_DEFAULT_VOICE = "alloy"


@dataclass
class OpenAISpeechModel:
    api_key: str
    base_url: str = _DEFAULT_BASE_URL
    name: str = "openai"

    async def generate(self, request: SpeechRequest) -> SpeechResult:
        model_id = request.provider_options.get("model", "tts-1")
        payload = {
            "model": model_id,
            "input": request.text,
            "voice": request.voice or _DEFAULT_VOICE,
            "response_format": request.output_format or "mp3",
        }

        async with httpx.AsyncClient(timeout=60) as client:
            response = await client.post(
                f"{self.base_url}/audio/speech",
                headers={
                    "Authorization": f"Bearer {self.api_key}",
                    "Content-Type": "application/json",
                },
                json=payload,
            )
            handle_error_response(response, "openai")

            return SpeechResult(
                audio=GeneratedAudioFile(
                    data=response.content,
                    media_type=f"audio/{payload['response_format']}",
                ),
                model_id=model_id,
                provider=self.name,
            )


def create_openai(
    *, api_key: str | None = None, base_url: str = _DEFAULT_BASE_URL
) -> OpenAISpeechModel:
    return OpenAISpeechModel(
        api_key=resolve_api_key(api_key=api_key, env_var="OPENAI_API_KEY", provider="openai"),
        base_url=base_url,
    )

File: packages/go/providers/openai/openai.go
// Package openai implements the OpenAI TTS provider.
package openai

import (
	"bytes"
	"context"
	"encoding/json"
	"io"
	"net/http"

	"github.com/jellypod-inc/speech-sdk/go/internal/httpx"
	"github.com/jellypod-inc/speech-sdk/go/speechtypes"
)

const (
	defaultBaseURL = "https://api.openai.com/v1"
	defaultVoice   = "alloy"
)

// Provider is the OpenAI speech provider.
type Provider struct {
	APIKey  string
	BaseURL string
	Client  *http.Client
}

// Option configures a Provider.
type Option func(*Provider)

// WithAPIKey sets the API key explicitly.
func WithAPIKey(key string) Option { return func(p *Provider) { p.APIKey = key } }

// WithBaseURL overrides the API base URL.
func WithBaseURL(url string) Option { return func(p *Provider) { p.BaseURL = url } }

// WithHTTPClient overrides the HTTP client used for requests.
func WithHTTPClient(c *http.Client) Option { return func(p *Provider) { p.Client = c } }

// New constructs a new OpenAI provider, resolving the API key from
// OPENAI_API_KEY if not supplied.
func New(opts ...Option) (*Provider, error) {
	p := &Provider{BaseURL: defaultBaseURL, Client: http.DefaultClient}
	for _, opt := range opts {
		opt(p)
	}
	key, err := httpx.ResolveAPIKey(p.APIKey, "OPENAI_API_KEY", "openai")
	if err != nil {
		return nil, err
	}
	p.APIKey = key
	return p, nil
}

// Name returns the provider identifier.
func (p *Provider) Name() string { return "openai" }

// Generate implements speechtypes.Provider.
func (p *Provider) Generate(ctx context.Context, req speechtypes.Request) (*speechtypes.Result, error) {
	modelID, _ := req.ProviderOptions["model"].(string)
	if modelID == "" {
		modelID = "tts-1"
	}
	voice := req.Voice
	if voice == "" {
		voice = defaultVoice
	}
	format := req.OutputFormat
	if format == "" {
		format = "mp3"
	}

	body, err := json.Marshal(map[string]any{
		"model":           modelID,
		"input":           req.Text,
		"voice":           voice,
		"response_format": format,
	})
	if err != nil {
		return nil, err
	}

	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, p.BaseURL+"/audio/speech", bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	httpReq.Header.Set("Authorization", "Bearer "+p.APIKey)
	httpReq.Header.Set("Content-Type", "application/json")

	resp, err := p.Client.Do(httpReq)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if err := httpx.HandleErrorResponse(resp, "openai"); err != nil {
		return nil, err
	}

	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	return &speechtypes.Result{
		Audio:    speechtypes.AudioFile{Data: data, MediaType: "audio/" + format},
		ModelID:  modelID,
		Provider: p.Name(),
	}, nil
}
